<<<<<<< HEAD
---
title: "Classification of regulatory sequences using machine learning techniques"
output:
  pdf_document: default
  pdf_notebook: default
bibliography: bib.bib
link-citations: yes
---
### Boris Shilov, Wim Thiels, Lucas Coppens, Zixuan Xie

\fontsize{12}{20}
# Introduction

\fontsize{12}{15}
\selectfont
Melanoma is an aggressive cancer with a high level of therapy-resistance due to its high degree of heterogeneity and plasticity. Despite the great efforts put into the field of melanoma treatment, resulting in great advances, many challenges remain. The major challange in fighting melanoma relapse continues to be the tumor heterogeneity, as melanoma comprises a wide variety of phenotypically distinct subpopulations of cancer cells. To be able to adress this issue therapeutically, the underlying mechanisms of heterogeneity need to be characterized. In this paper, a distinction between two major types of melanoma cells is made: invasive and proliferative. [@Verfaillie; @Shannan]


Transcriptional reprogramming of melanoma cells in proliferative state into melanoma cells with invasive characteristics is a critical event at the origin of metastatic spreading of melanoma. Invasive cells have acquired the ability to migrate to other tissues, enter the bloodstream and therefore lie at the basis of the metastatic spreading of cancer in the body. While the transitional mechanisms from proliferative to invasive cancer cell are yet to be characterized more extensively, it is sure that one event lies at the basis of this transition: the transcriptional reprogramming of the cell. Studying the involved genes and regulatory elements using various bioinformatics approaches is therefore a hot topic in the area of melanoma research. Decoding the regulatory landscape could result into the ability to push melanoma cells towards a different cell state, which would be an interesting target from a therapeutic point of view. [@Verfaillie]


Transcriptomic, open chromatin and histone modification maps of melanoma cultures were constructed, revealing thousands of active cis-regulatory regions, both for proliferative and invasive cells. [@Verfaillie] It should be possible to discern which cis-regulatory regions are useful for the classification of cell states in melanoma samples if such states are truly distinct in terms of regulatory landscape. The aim of the project was the construction of classifiers predicting whether a regulatory region would be active in proliferative or invasive cell states. Another useful insight that would be gained from these classifiers, is which regulatory regions are the most significant for distinguishing between cell states, thus giving information about the underlying mechanisms and the critical genes and regulatory elements involved in cancer cell state transitions. 


Two distinct machine learning techniques were used for the creation of such classifiers, namely the random forests ensemble method and deep learning, with the use of convolutional neural networks. Both models were trained on the same training set, which comprises the dataset of active cis-regulatory regions, mentioned hereabove. In this paper, both methods are described and their results are evaluated and compared.

# Methods

# Results and discussion

## Deep learning
### A deep learning approach to motif finding
As shown in Fig1 a deep neural network takes the raw DNA sequences as input. No manually crafted dataset of features is needed, which is one of the major advantages of using a deep learning approach.   After transformation to a 1-hot representation, the DNA sequences can be inputted into a first convolutional layer.   Here a number of kernels, acting as motif scanners, will go over the sequences.   Their width equals 4 (number of nucleotides) and their length corresponds to the length of the motifs we try to extract.   Likewise, the number of kernels will correspond to the maximum number of motifs we will be able to find.   This first convolutional layer is then followed by consecutive convolutional layers, each time transforming the input data into more higher level features.    At certain points drop out layers and batch normalization are used to reduce the number of parameters, and as a counterweight to overfitting.  At the end, these extracted featured are combined through some dense layers and eventually passed through a softmax layer which outputs a probability distribution over the different classes (in our case binary , P or I).   During training, the network will adjust its weights so that classification is improved.   This will lead to an adjustment of the weights in the kernels of the first layer to match the characteristic differences between the two types of DNA sequences (=the motifs).
In addition, some form of recurrent layers can be incorporated as well.  A popular choice are Long Short Term Memory networks (LSTMs). These recurrent neural layers can pick up on sequential information and are therefore better suited to capture the context of the motifs (motif syntax).

![Figure 1: Overview of a deep learning set up for motif detection.  From Min, Xu et al (2017)][fig1]

[fig1]: https://github.com/muscovitebob/EnhancerMachineLearning/tree/master/report/pics/Overview set up deep learning for motif detection.png "Overview set up deep learning for motif detection"

### Comparing 2 deep learning architectures
=======
# References

>>>>>>> 5a54e5c5a0b9df7b9baf74f7184f269635e0ee86
